admin/services/redis-cache.service.js

```diff
@@ -0,0 +1,213 @@
+const Redis = require('ioredis')
+const config = require('../config')
+const azureQueueService = require('./azure-queue.service')
+const queueNameService = require('./queue-name-service')
+
+const { REDIS_CACHING } = config
+
+let redisConfig = {
+  port: config.Redis.Port,
+  host: config.Redis.Host
+}
+if (config.Redis.Key) {
+  redisConfig.password = config.Redis.Key
+}
+if (config.Redis.useTLS) {
+  redisConfig.tls = { host: config.Redis.Host }
+}
+
+const redis = new Redis(redisConfig)
+
+const redisCacheService = {}
+
+redisCacheService.affectedTables = {
+  'checkWindow.sqlFindActiveCheckWindow': ['checkWindow'],
+  'group.sqlFindGroups': ['group', 'pupilGroup'],
+  'schoolData.sqlFindOneById': ['sce', 'school']
+}
+
+redisCacheService.storedProceduresAffectedTables = {
+  'spUpsertSceSchools': ['sce']
+}
+
+/**
+ * Constructs a full Redis key in the format `serviceName.methodName_affectedTable1-affectedTable2`
+ * @param serviceKey - the prefix of the key in the format `serviceName.methodName`
+ * @returns {String}
+ */
+redisCacheService.getFullKey = serviceKey => {
+  if (!serviceKey || !REDIS_CACHING) {
+    return false
+  }
+  let affectedKey = serviceKey
+  const keyParts = serviceKey.split('.')
+  if (keyParts.length > 2) {
+    // the key includes an identifier e.g. school_id
+    affectedKey = `${keyParts[0]}.${keyParts[1]}`
+  }
+  const tables = redisCacheService.affectedTables[affectedKey]
+  if (tables) {
+    serviceKey = `${serviceKey}_${tables.join('-')}`
+  }
+  return serviceKey
+}
+
+/**
+ * Returns the data from a Redis key entry
+ * @param redisKey - the full Redis key to get
+ * @returns {String}
+ */
+redisCacheService.get = redisKey => {
+  return new Promise((resolve, reject) => {
+    redis.get(redisKey, (err, result) => {
```

Comment:
It would return a false result, allowing the query to proceed, because Redis errored. But to be thorough, I’ve added `reject` and a try/catch.

1. Before executing a SQL Query which is set to be cached in Redis, it looks for an existing cache.
2. If it can’t find it or there is an error, it returns false, which executes the SQL query (the result set is then cached)
3. If it can find it, it returns the cache

I’ve tweaked #2 to cause a `reject` error on error. So “no result” and error are treated slightly differently.